frontend/hub.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"bytes"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+type Client struct {
+	hub *Hub
+
+	// The websocket connection.
+	conn *websocket.Conn
+
+	lock sync.Mutex
+
+	// Buffered channel of outbound messages.
+	send chan []byte
+}
+
+const (
+	pongWait = 5 * time.Second
+)
+
+type Hub struct {
+	clients map[*Client]bool
+
+	broadcast chan []byte
+
+	register chan *Client
+
+	unregister chan *Client
+}
+
+var hub = &Hub{
+	broadcast:  make(chan []byte),
+	register:   make(chan *Client),
+	unregister: make(chan *Client),
+	clients:    make(map[*Client]bool),
+}
+
+func (h *Hub) run() {
+	for {
+		select {
+		case client := <-h.register:
+			h.clients[client] = true
+		case client := <-h.unregister:
+			if _, ok := h.clients[client]; ok {
+				delete(h.clients, client)
+				close(client.send)
+			}
+		case message := <-h.broadcast:
+			for client := range h.clients {
+				select {
+				case client.send <- message:
+				default:
+					close(client.send)
+					delete(h.clients, client)
```

Comment:
if the routine sending msgs to browser over is slower than the h.broadcast ch, then you are disconnecting the client. just drop missed messages.